controllers/agentmachine_controller.go

```diff
@@ -544,8 +540,42 @@ func getAddresses(foundAgent *aiv1beta1.Agent) []clusterv1.MachineAddress {
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *AgentMachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (r *AgentMachineReconciler) SetupWithManager(mgr ctrl.Manager, agentNamespace string) error {
+	mapAgentToAgentMachine := func(agent client.Object) []reconcile.Request {
+		log := r.Log.WithFields(
+			logrus.Fields{
+				"agent":           agent.GetName(),
+				"agent_namespace": agent.GetNamespace(),
+			})
+
+		namespace, ok := agent.GetAnnotations()[AgentMachineRefNamespace]
+		if !ok {
+			return make([]reconcile.Request, 0)
+		}
+
+		amList := &capiproviderv1alpha1.AgentMachineList{}
+		opts := &client.ListOptions{
+			Namespace: namespace,
+		}
+		if err := r.List(context.Background(), amList, opts); err != nil {
+			log.Debugf("failed to list agent machines")
+			return []reconcile.Request{}
+		}
+
+		reply := make([]reconcile.Request, 0, len(amList.Items))
+		for _, agentMachine := range amList.Items {
+			if agentMachine.Status.AgentRef.Namespace == agent.GetNamespace() && agentMachine.Status.AgentRef.Name == agent.GetName() {
+				reply = append(reply, reconcile.Request{NamespacedName: types.NamespacedName{
+					Namespace: agentMachine.Namespace,
+					Name:      agentMachine.Name,
+				}})
+			}
+		}
+		return reply
```

Comment:
Yes, you're right, there should be at most one.  The problem is how we get from an agent to its agentMachine.  I added the namespace in an annotation which was easy because both have at most 63 characters.  So at least we narrow down the scope of the list.  However, the name of the agentMachine can be up to 253 characters so it doesn't fit in an annotation/label.  I didn't find any other property on the agent to put this reference, and I don't know if we should add one.  We could theoretically compress the string and/or split it among multiple annotations, but this seems overly complicated.